extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -162,49 +162,27 @@ public void close() {
         if (log.isDebugEnabled()) {
             log.debug().append("Closing context: ").append(ctxStr()).endl();
         }
-        releaseOutstanding();
-    }
-
-    /**
-     * Release all outstanding requests associated with this context. Eventually, the request will be canceled when the
-     * objects are garbage collected.
-     */
-    private void releaseOutstanding() {
-        Arrays.fill(localCache, null);
     }
 
     // --------------------------------------------------------------------------------------------------
 
     @Nullable
-    private S3Request getRequestFromLocalCache(final long fragmentIndex) {
-        return getRequestFromLocalCache(fragmentIndex, cacheIndex(fragmentIndex));
-    }
-
-    @Nullable
-    private S3Request getRequestFromLocalCache(final long fragmentIndex, final int cacheIdx) {
-        if (localCache[cacheIdx] != null && localCache[cacheIdx].request.isFragment(fragmentIndex)) {
-            return localCache[cacheIdx].request;
+    private S3Request.AcquiredRequest getRequestFromSharedCache(final long fragmentIndex) {
+        final S3Request.AcquiredRequest cachedRequest = sharedCache.getRequest(uri, fragmentIndex);
+        if (cachedRequest == null) {
+            return null;
         }
-        return null;
+        // Send the request, if not sent already. The following method is idempotent, so we always call it.
+        cachedRequest.request.sendRequest();
```

Comment:
I understand the comment; but do we ever expect this to be of practical concern? I'm assuming in the case of a getRequest, it must be the case that something else already did getOrCreateRequest + sendRequest?

Reply:
This was a race condition I had detected when I added the shared cache logic.

Basically, the sendRequest method has a synchronized block at the end of which the primary `consumerFuture` gets populated. So if say thread 1 comes in and is half way through sendRequest and thread 2 doesn't call sendRequest, then it could find `consumerFuture` as null which leads to a NPE at the time of filling. So I force each thread to always call sendRequest so that they all block till the request is fully sent and consumerFuture is populated.

Reply:
I tweaked it up a little bit as per your suggestion. Let me know your thoughts.